### DISCUSSION TITLE
Working with set of parralel tasks

### DISCUSSION THREAD

**Discussion by JekRock** [statelyai/xstate#3527] [Q&A] [answered]
What is the “right” way to model the following scenario with `xstate`?

I have a set of tasks. Every task is an instance of the same state machine. There is a “parent” state machine that spawns this set of tasks in parallel as a part of one of its states.

I don't think I understand how to model a set of parallel **instances** of the same state machine.

As an analogy, we can model a `Promise` machine. But how can we model a `Promise.allSettled` machine, for example?

Thank you in advance.

**Comment by JekRock** [comment]
I was thinking it must be something related to actors. Let me know if I'm moving in a right direction.

**Comment by davidkpiano** [comment, marked as answer]
You can actually use `Promise.allSettled`, etc. directly:

```js
invoke: {
  src: (ctx, e) => Promise.allSettled(promises),
  onDone: { ... }
}
```

You can model a machine as a promise by using the `waitFor` helper:

```js
const service = interpret(machine).start();

const servicePromise = waitFor(service, state => state.matches('someState'));
```

**Reply by JekRock** [reply]
Thank you, it seems like it's exactly what I needed!
So, no need to `spawn` machines (as I don't need them to communicate), and they can just be started from the parent machine and wrapped in promises using `waitFor` :thumbsup:

**Reply by JekRock** [reply]
And thank you for the great library!
I've just started using it for modeling business processes in service. The service's execution flow seems much more transparent and less ambiguous now. State machines give confidence in state correctness.
